third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerAref.cpp

```diff
@@ -465,7 +465,40 @@ void insertArriveBarrier(Location loc, ArrayRef<AsyncOp> asyncOps,
 void rewritePutExitOp(ArefPutExitOp op, PatternRewriter &rewriter,
                       ArefValue arefVal) {
   auto loc = op->getLoc();
+  auto stageCluster = getStageCluster(op);
+  auto asyncKinds = castAsyncOpAttrs(op.getAsyncOps());
   rewriter.setInsertionPointAfter(op);
+
+  bool needFence = [&]() {
+    bool isGenericProxy = llvm::any_of(
+        asyncKinds, [](AsyncOp kind) { return kind == AsyncOp::NONE; });
+    if (!isGenericProxy) {
+      return false;
+    }
+    auto tmem = TensorMemorySpaceAttr::get(op.getContext());
+    auto arefType = cast<ArefType>(op.getAref().getType());
+    auto arefBufType = cast<MemDescType>(arefType.getBaseType()[0]);
+    if (arefBufType.getMemorySpace() == tmem) {
```

Comment:
nit: we may want to verify that all buffers have tmem type not just the first one, although we don't have use cases like that today. A comment would also be fine for posterity.